docs/man/man3/DS18B20.3
.TH "DS18B20" 3 "Thu May 25 2017" "Version 0.2" "SensorNode" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DS18B20 \- \fBDS18B20\fP class definition\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <DS18B20\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBReadMeasurement\fP (char *string)"
.br
.RI "\fITakes a Measurement per sensor on the bus\&. \fP"
.ti -1c
.RI "void \fBReadMeasurement2\fP (char *Macstring, char *tempStr)"
.br
.RI "\fITakes a Measurement per sensor on the bus\&. \fP"
.ti -1c
.RI "\fBDS18B20\fP ()"
.br
.RI "\fI\fBDS18B20\fP standard constructor\&. \fP"
.ti -1c
.RI "\fBDS18B20\fP (volatile uint8_t &Wire_PORT, volatile uint8_t &Wire_DDR, volatile uint8_t &Wire_PIN, uint8_t Wire_DQ)"
.br
.RI "\fI\fBDS18B20\fP constructor\&. \fP"
.ti -1c
.RI "void \fBGetMACString\fP (char *MACStringOut)"
.br
.RI "\fIGet MAC addresses from MACString\&. \fP"
.ti -1c
.RI "void \fBGetSensorStringXML\fP (char *string)"
.br
.RI "\fIGet xml style sensor string\&. \fP"
.ti -1c
.RI "void \fBGetTemperature\fP (char *MACstring, char *tempString)"
.br
.RI "\fIGet temperature values in one string an the MAC addresses in a seperate one\&. \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBStartMeasurement\fP (void)"
.br
.RI "\fISensor Initialization\&. \fP"
.ti -1c
.RI "uint8_t \fBReset\fP ()"
.br
.ti -1c
.RI "uint8_t \fBBitIO\fP (uint8_t bit)"
.br
.RI "\fIBitIO\&. \fP"
.ti -1c
.RI "uint8_t \fBWriteByte\fP (uint8_t byte)"
.br
.ti -1c
.RI "uint8_t \fBReadByte\fP (void)"
.br
.ti -1c
.RI "unsigned char \fBRomSearch\fP (unsigned char diff, unsigned char *id)"
.br
.ti -1c
.RI "void \fBCommand\fP (unsigned char command, unsigned char *id)"
.br
.ti -1c
.RI "void \fBPIN_INPUT_MODE\fP ()"
.br
.ti -1c
.RI "void \fBPIN_OUTPUT_MODE\fP ()"
.br
.ti -1c
.RI "void \fBPIN_LOW\fP ()"
.br
.ti -1c
.RI "void \fBPIN_HIGH\fP ()"
.br
.ti -1c
.RI "\fB__attribute__\fP ((gnu_inline)) void delay_us(uint16_t delay)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "volatile uint8_t * \fBWIRE_PORT\fP"
.br
.RI "\fIThermometer data port\&. \fP"
.ti -1c
.RI "volatile uint8_t * \fBWIRE_DDR\fP"
.br
.RI "\fIThermometer data port\&. \fP"
.ti -1c
.RI "volatile uint8_t * \fBWIRE_PIN\fP"
.br
.RI "\fIThermometer data port\&. \fP"
.ti -1c
.RI "uint8_t \fBWIRE_DQ\fP"
.br
.RI "\fIThermometer data line\&. \fP"
.ti -1c
.RI "char \fBMACString\fP [120]"
.br
.RI "\fIMAC string of the sensor\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
\fBDS18B20\fP class definition\&. 

Defines methods for accessing the \fBDS18B20\fP interface 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "DS18B20::DS18B20 ()"

.PP
\fBDS18B20\fP standard constructor\&. new \fBDS18B20\fP object and initializes variables 
.SS "DS18B20::DS18B20 (volatile uint8_t & Wire_PORT, volatile uint8_t & Wire_DDR, volatile uint8_t & Wire_PIN, uint8_t Wire_DQ)"

.PP
\fBDS18B20\fP constructor\&. 
.PP
\fBParameters:\fP
.RS 4
\fIWire_PORT\fP 
.br
\fIWire_DDR\fP 
.br
\fIWire_PIN\fP 
.br
\fIWire_DQ\fP Data line 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "DS18B20::__attribute__ ((gnu_inline))\fC [inline]\fP, \fC [private]\fP"

.SS "uint8_t DS18B20::BitIO (uint8_t bit)\fC [private]\fP"

.PP
BitIO\&. 
.PP
\fBParameters:\fP
.RS 4
\fIbit\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
.RE
.PP

.SS "void DS18B20::Command (unsigned char command, unsigned char * id)\fC [private]\fP"

.SS "void DS18B20::GetMACString (char * MACStringOut)"

.PP
Get MAC addresses from MACString\&. 
.PP
\fBParameters:\fP
.RS 4
\fIMACStringOut\fP Character array of MAC addresses 
.RE
.PP

.SS "void DS18B20::GetSensorStringXML (char * string)"

.PP
Get xml style sensor string\&. Puts out one string with the temperature values 
.PP
\fBParameters:\fP
.RS 4
\fIstring\fP XML style sensor string 
.RE
.PP

.SS "void DS18B20::GetTemperature (char * MACstring, char * tempString)"

.PP
Get temperature values in one string an the MAC addresses in a seperate one\&. 
.PP
\fBParameters:\fP
.RS 4
\fIMACstring\fP MAC addresses of each sensor on the bus 
.br
\fItempString\fP Temperature values in one string 
.RE
.PP

.SS "void DS18B20::PIN_HIGH ()\fC [inline]\fP, \fC [private]\fP"

.SS "void DS18B20::PIN_INPUT_MODE ()\fC [inline]\fP, \fC [private]\fP"

.SS "void DS18B20::PIN_LOW ()\fC [inline]\fP, \fC [private]\fP"

.SS "void DS18B20::PIN_OUTPUT_MODE ()\fC [inline]\fP, \fC [private]\fP"

.SS "uint8_t DS18B20::ReadByte (void)\fC [private]\fP"

.SS "void DS18B20::ReadMeasurement (char * string)"

.PP
Takes a Measurement per sensor on the bus\&. 
.PP
\fBParameters:\fP
.RS 4
\fIstring\fP String with MAC addresses and temperature values 
.RE
.PP

.SS "void DS18B20::ReadMeasurement2 (char * Macstring, char * tempStr)"

.PP
Takes a Measurement per sensor on the bus\&. 
.PP
\fBParameters:\fP
.RS 4
\fIMacstring\fP MAC addresses of each sensor on the bus 
.br
\fItempStr\fP Temperature values in one string 
.RE
.PP

.SS "uint8_t DS18B20::Reset ()\fC [private]\fP"

.SS "unsigned char DS18B20::RomSearch (unsigned char diff, unsigned char * id)\fC [private]\fP"

.SS "void DS18B20::StartMeasurement (void)\fC [private]\fP"

.PP
Sensor Initialization\&. Starts 
.SS "uint8_t DS18B20::WriteByte (uint8_t byte)\fC [private]\fP"

.SH "Member Data Documentation"
.PP 
.SS "char DS18B20::MACString[120]\fC [private]\fP"

.PP
MAC string of the sensor\&. 
.SS "volatile uint8_t* DS18B20::WIRE_DDR\fC [private]\fP"

.PP
Thermometer data port\&. 
.SS "uint8_t DS18B20::WIRE_DQ\fC [private]\fP"

.PP
Thermometer data line\&. 
.SS "volatile uint8_t* DS18B20::WIRE_PIN\fC [private]\fP"

.PP
Thermometer data port\&. 
.SS "volatile uint8_t* DS18B20::WIRE_PORT\fC [private]\fP"

.PP
Thermometer data port\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for SensorNode from the source code\&.


src/Sensors/DS18B20.cpp
/**
****************************************************************************
* Project: SensorNode
*
* DS18B20.cpp
*
* Created on: Nov 12, 2016
*
* Author: Matthias Minx
*
* Revision: 0.2
*
*
****************************************************************************/
/** \file DS18B20.cpp
 *
 */

#include "DS18B20.h"
#include <util/delay.h>
#include <inttypes.h>
#include <avr/interrupt.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

//#include "../IO/USART.h"

namespace Sensors {

DS18B20::DS18B20() {


	WIRE_PORT = &PORTB;

	WIRE_DDR = &DDRB;

	WIRE_PIN = &PINB;

	WIRE_DQ  = PB0;




//u = USART();


}


DS18B20::DS18B20(volatile uint8_t &Wire_PORT, volatile uint8_t &Wire_DDR, volatile uint8_t &Wire_PIN,uint8_t Wire_DQ){



	WIRE_PORT = &Wire_PORT;

	WIRE_DDR = &Wire_DDR;

	WIRE_PIN = &Wire_PIN;

	WIRE_DQ  = Wire_DQ;

	// u = USART();





}

void DS18B20::StartMeasurement( void ){
	 if( *WIRE_PIN & 1<< WIRE_DQ){
		 Command( THERM_CMD_CONVERTTEMP, NULL );
		*WIRE_PORT |= 1<< WIRE_DQ;
	    *WIRE_DDR |= 1<< WIRE_DQ;			// parasite power on

	  }else{
	    ;
	  }
}


void DS18B20::ReadMeasurement( char* tempStr ){

	unsigned char id[8], diff;

	sprintf(tempStr,"%s",(char*)"");
	sprintf(MACString,"%s",(char*)"");
	//unsigned char i;
	unsigned int temp;

	for( diff = SEARCH_FIRST; diff != LAST_DEVICE; ){
		diff = RomSearch( diff, id );

	    if( diff == PRESENCE_ERR ){
	    	sprintf(tempStr, "Error 0x00|" ); // No sensor found
	    	break;
	    }
	    if( diff == DATA_ERR ){
	    	sprintf(tempStr, "Error 0x10|" ); // Bus Error
	    	break;
	    }
	    if( id[0] == 0x28 || id[0] == 0x10 ){	// temperature sensor
	    	sprintf( MACString, "%s%02X:%02X:%02X:%02X:%02X:%02X:%02X:%02X|",MACString, id[0], id[1], id[2], id[3], id[4], id[5], id[6], id[7] );
	    	WriteByte( THERM_CMD_RSCRATCHPAD );			// read command
	    	temp = ReadByte();			// low byte
	    	temp |= (unsigned int)ReadByte() << 8;		// high byte
	    	// temp-=0x20;
	    	if( id[0] == 0x10 )			// 9 -> 12 bit
	    		temp <<= 3;
	    	//   sprintf( string, "  %04X = ", temp );	// hex value

	    	sprintf( tempStr, "%s%d.%01d°C|",tempStr, temp >> 4, (temp << 12) / 6553 ); // 0.1C

	    }

	  }


}

void DS18B20::ReadMeasurement2( char* Macstring, char* tempStr ){

	unsigned char id[8], diff;
	sprintf(Macstring,(char*)" ");
	sprintf(tempStr,(char*)" ");
	//unsigned char i;
	unsigned int temp;

	for( diff = SEARCH_FIRST; diff != LAST_DEVICE; ){
		diff = RomSearch( diff, id );

	    if( diff == PRESENCE_ERR ){
	    	sprintf(Macstring, "No Sensor found." );
	    	break;
	    }
	    if( diff == DATA_ERR ){
	    	sprintf(Macstring, "Bus Error." );
	    	break;
	    }
	    if( id[0] == 0x28 || id[0] == 0x10 ){	// temperature sensor
	    	sprintf( Macstring, "%s%02X:%02X:%02X:%02X:%02X:%02X:%02X:%02X|",Macstring, id[0], id[1], id[2], id[3], id[4], id[5], id[6], id[7] );
	    	WriteByte( THERM_CMD_RSCRATCHPAD );			// read command
	    	temp = ReadByte();			// low byte
	    	temp |= (unsigned int)ReadByte() << 8;		// high byte
	    	// temp-=0x20;
	    	if( id[0] == 0x10 )			// 9 -> 12 bit
	    		temp <<= 3;
	    	//   sprintf( string, "  %04X = ", temp );	// hex value

	    	sprintf( tempStr, "%s%d.%01d°C|",tempStr, temp >> 4, (temp << 12) / 6553 ); // 0.1C

	    }

	  }


}

uint8_t DS18B20::Reset(){
	uint8_t i;
	 //Pull line low and wait for 480uS
	PIN_LOW();
	PIN_OUTPUT_MODE();
	delay_us(us(480));
	cli();
	//Release line and wait for 60uS
	PIN_INPUT_MODE();
	delay_us(us(60));
	//Store line value and wait until the completion of 480uS period
	i=(*WIRE_PIN & (1<<WIRE_DQ));
	sei();

	delay_us(us(420));
	//Return the value read from the presence pulse (0=OK, 1=WRONG)
	if( (*WIRE_PIN & (1<<WIRE_DQ)) == 0 )		// short circuit
	 i = 1;


	return (i);

}
uint8_t DS18B20::BitIO(uint8_t bit){
	 //Pull line low for 1uS
		 cli();
		 PIN_LOW();
		 PIN_OUTPUT_MODE();
		 delay_us(us(1));

		 //If we want to write 1, release the line (if not will keep low)
		 if(bit) PIN_INPUT_MODE();
		 delay_us(15);
		 if( (*WIRE_PIN & (1<<WIRE_DQ)) == 0 )
			 bit=0;
		 //Wait for 60uS and release the line
		 delay_us(us(60));
		 PIN_INPUT_MODE();
		sei();
	return (bit);
}
uint8_t DS18B20::WriteByte(uint8_t byte){
	uint8_t i=8, j;

	 	 do{
	 		//Write actual bit and shift one position right to make the next bit ready
	 		j = BitIO(byte&1);
	 		 byte>>=1;
	 		 if(j)
	 			byte |= 0x80;

	 	 }while(--i);

	return (byte);
}
uint8_t DS18B20::ReadByte(void){
	 return (WriteByte( 0xFF ));
}
unsigned char DS18B20::RomSearch(unsigned char diff, unsigned char *id){
	unsigned char i, j, next_diff;
	unsigned char b;

	  if( Reset() )
	    return (PRESENCE_ERR);			// error, no device found
	  WriteByte( SEARCH_ROM );			// ROM search command
	  next_diff = LAST_DEVICE;			// unchanged on last device
	  i = 8 * 8;					// 8 bytes
	  do{
	    j = 8;					// 8 bits
	    do{
	      b = BitIO( 1 );			// read bit
	      if( BitIO( 1 ) ){			// read complement bit
		if( b )					// 11
		  return (DATA_ERR);			// data error
	      }else{
		if( !b ){				// 00 = 2 devices
		  if( diff > i ||
		    ((*id & 1) && diff != i) ){
		    b = 1;				// now 1
		    next_diff = i;			// next pass 0
		  }
		}
	      }
	      BitIO( b );     			// write bit
	      *id >>= 1;
	      if( b )					// store bit
		*id |= 0x80;
	      i--;
	    }while( --j );
	    id++;					// next byte
	  }while( i );
	  return (next_diff);				// to continue search


}
void DS18B20::Command( unsigned char command, unsigned char *id ){
	unsigned char i;
	Reset();
	  if( id ){
		  WriteByte( MATCH_ROM );			// to a single device
	    i = 8;
	    do{
	    	WriteByte( *id );
	      id++;
	    }while( --i );
	  }else{
		  WriteByte( SKIP_ROM );			// to all devices
	  }
	  WriteByte( command );
}

void DS18B20::GetMACString(char* MACStringOut){
	sprintf(MACStringOut,"%s",MACString);
}
/*
void DS18B20::GetTemperatureString(char* TemperatureString){


}
*/
void DS18B20::GetSensorStringXML(char* string){

	//char str[255];
	StartMeasurement();
//	ReadMeasurement(str);
//_delay_ms(2000);
	ReadMeasurement(string);

}
void DS18B20::GetTemperature(char* MACstring,char* tempString){

	//char str[255];
	StartMeasurement();
//	ReadMeasurement(str);
//_delay_ms(2000);
	ReadMeasurement2( MACstring, tempString);

}

}
